Complete rework using bleak

dhalbert opened the pull request:
Bleak re-implementation of `_bleio`.

Notes:
- bleak only works as a central.
- Tested on WIndows 10 and Linux with various devices, including Nordic UART, heart rate monitor, iBBQ, and Magic Light Bulb.
- **Doesn't yet work on MacOS** due to event-loop vs. threads problem in bleak: https://github.com/hbldh/bleak/issues/206, which is under active discussion.
- **Doesn't yet work with Adafruit Services** due to bleak's assumption that all characteristics in all services have unique UUIDs. Adafruit Services have several shared characteristic ID's. This is being worked on: https://github.com/hbldh/bleak/issues/139, https://github.com/hbldh/bleak/pull/167.
- Underlying implementations (dotnet, bluez) may need resetting in various ways if an error occurs or between some uses. Bluez may need to be directed to forget devices when those devices change what they advertise or provide.
- Windows 10 release 2004 (very recent) may need the Bluetooth device to be disabled and re-enabled to remove the "broken device" notation in Device Manager.
- Must use Python 3.7 or 3.8 or Linux.
- Must use Python 3.7 only on Windows, because the underlying dotnet library is not yet compatible with 3.8.

dhalbert commented on _bleio/adapter_.py:
```diff
@@ -0,0 +1,222 @@
+        :returns: an iterable of `_bleio.ScanEntry` objects
+        :rtype: iterable"""
+
+        scanner = BleakScanner(loop=self._bleak_loop)
```
Yes, bleak uses bluez. I saw that you used `hcitool`, but `hcitool` is part of the bluez source tree, so is it not also using bluez? Also I've read that it's deprecated. If there's filtering going on, it might be some defaults that bleak set on bluez.

What filtering specifically did you see? Could we get around it by doing very short scanning intervals, if the filtering is during one scanning period? Already I am using 0.25 second scanning periods to simulate real-time scanning results.

tannewt replied:
`hcitool` has raw hci access to the ble module. When I did this, I dug all the way down and found that the linux kernel always deduplicates scans by address iirc. There is no setting otherwise. https://marc.info/?l=linux-bluetooth&m=158225950522806&w=2

Whether 0.25 seconds is enough depends on the configuration of the broadcaster. I was testing this with the `missed-message-count` of the BroadcastNet bridge. It tracks how many sequence numbers are missed from the broadcasts and is one way to get a feel for how many are missed.

dhalbert replied:
I'll open an issue about this, for sure, to be investigated and fixed. I'd also like to lobby for them to fix that thing in bluez, since one should be able to turn off the deduplication. I don't want to defer releasing the library until fixing it, though, since it has a bunch of use cases right now that aren't advertising-only.

dhalbert replied:
I don't know that the BLE adapter is always accessible via hcitool, though it probably usually is. I don't know if it supports any other adapters besides via HCI.

tannewt replied:
I'm pretty sure all BLE adapters are HCI over one transport or another. Is it a ton of work to use hcitool for scanning on Linux? I'm a bit frustrated that I did a bunch of work on this and it's now being replaced by code that won't work as well.

dhalbert replied:
I'm sorry, I can just use your code. I didn't see the conversation you had on the bluez email list until just now; I was only looking at the kernel flag.

Whether `hcitool` and other tools were deprecated and removed or not is really confusing. The Arch Linux people say they were, (as of bluez 5.44?), but they are still in the bluez.org download packages. So I'll include your code but fallback to the current implementation.

dhalbert replied:
Just to be clear, I am not using ArchLinux; that's just where I found the discussion. Ubuntu 20.04 uses bluez 5.43.

tannewt replied:
Falling back would be awesome. My main platform of concern is Raspbian.

dhalbert replied:
I added and adapted your original code. I added a check for `hcitool` having the correct permissions by attempting to do an `hcitool cmd <noop>`, which will fail without the permissions.

tannewt commented on _bleio/adapter_.py:
```diff
@@ -0,0 +1,222 @@
+
+    @property
+    def address(self) -> Address:
+        # bleak has no API for the address yet.
```
I found a way to do it on Linux: https://github.com/adafruit/Adafruit_Blinka_bleio/blob/master/_bleio.py#L419-L423

dhalbert replied:
I am fetching the address using `bluetoothctl list` command instead. It's a bit simpler and the command clearly marks which adapter is the default, if there are multiple ones.

tannewt commented on _bleio/characteristic_buffer.py:
```diff
@@ -0,0 +1,140 @@
+        for b in data:
+            try:
+                self._queue.put_nowait(b)
+            except queue.Full:
```
Can you check full with an if statement or are you worried about a race condition?

dhalbert replied:
This is the recommended way, I believe, and is the least likely to cause a race.

tannewt commented on _bleio/characteristic_buffer.py:
```diff
@@ -0,0 +1,140 @@
+        while len(b) < nbytes and time.time() < end:
+            try:
+                b.append(self._queue.get_nowait())
+            except queue.Empty:
```
Here too, can you change this exception to a test up front?

dhalbert replied:
ditto

tannewt commented on _bleio/characteristic_buffer.py:
```diff
@@ -0,0 +1,140 @@
+        :rtype: int or None (on a non-blocking error)
+        """
+        bytes_read = self.read(len(buf))
+        buf[0 : len(bytes_read)] = bytes_read
```
Please swap these and call readinto from read. Let's not always do a bytearray allocation and then a copy.

I know this is on full computers but I worry folks will pick up on this pattern and use it on smaller devices.

tannewt commented on _bleio/characteristic_buffer.py:
```diff
@@ -0,0 +1,140 @@
+
+    def deinit(self,) -> None:
+        """Disable permanently."""
+        self._characteristic.add_notify_callback(self._notify_callback)
```
Delete the callback here.

dhalbert replied:
Yep, this is wrong.

tannewt commented on _bleio/connection.py:
```diff
@@ -0,0 +1,188 @@
+        return conn
+
+    @property
+    def bleak_client(self):
```
Why make this public? Doing so means that code written for RPi won't work in CircuitPython.

tannewt replied:
(People can just access `_bleak_client` if they really want to.)

tannewt commented on _bleio/packet_buffer.py:
```diff
@@ -0,0 +1,119 @@
+          or a remote Characteristic in a remote Service that a Central has connected to.
+        :param int buffer_size: Size of ring buffer (in packets of the Characteristic's maximum
+          length) that stores incoming packets coming from the peer."""
+        self._queue = queue.Queue(buffer_size)
```
Don't you need two queues? One incoming and one outgoing?

dhalbert replied:
@tannewt and I ended up talking about this. I can't do an outgoing queue because I can't tell that there's a pending packet still waiting to be sent. So all the packets are just sent one at at time: no concatenation is possible.

tannewt commented on _bleio/packet_buffer.py:
```diff
@@ -0,0 +1,119 @@
+        """
+        # TODO Figure out if we can accumulate intermediate results.
+        packet = header + data if header else data
+        self._queue.put_nowait(packet)
```
Shouldn't this actually send?

dhalbert replied:
Yes, this is wrong. Fixed.

tannewt commented on _bleio/scan_entry.py:
```diff
@@ -0,0 +1,155 @@
+    @property
+    def connectable(self):
+        """This is unknown on most Blinka implementations. Always returns ``None``."""
+        return None
```
Any concern about None evaluating to False? Most scans are probably connectable.

dhalbert replied:
This is hard. Maybe it could be a better guess based on what's in the advertisement.

tannewt replied:
True. It's likely if any service related ADTs are included.

dhalbert replied:
Now I check for services in the advertisement and guess that it means it's connectable.

tannewt commented on _bleio/service.py:
```diff
@@ -0,0 +1,116 @@
+        return service
+
+    @property
+    def bleak_service(self):
```
Again, why is this public? (Same question for `from_bleak`.)

dhalbert replied:
It needed to be called from other classes, it's not internal to `Service only`. Basically it is like a C++ "friend" method. If I underscore it, I'll have to pylint-disable a lot of places. I guess that's the other choice, so I'll do taht.

tannewt commented on requirements.txt:
```diff
@@ -1 +1,3 @@
+Adafruit-Blinka
 bleak
+janus
```
What is janus and why do you need it?

dhalbert replied:
I don't need it any more; I forgot to remove it. It is a queue that you can use both from the sync and async side. Originally I was using it to pass the scanning results back, but I restructured things and it became unnecessary.

https://github.com/aio-libs/janus

dhalbert commented:
@tannewt Notable changes since last review:

- Added `hcitool` scanning. Falls back to `bleak` scanning if `hcitool` is not present or does not have correct capabilities.
- Use `bluetoothctl` to get MAC address.
- Rename a bunch of library-internal methods and properties to include a leading underscore, even across classes, to make it clear these are not for external use.
-Fixed bugs.

@ladyada Still some flakiness on Windows 10 version 2004 or later. Sometimes it works fine, sometimes it mysteriously doesn't. I saw this before you tested, and then the problem went away, and I thought it was a red herring. The issue is that notification callbacks sometimes don't happen. It seems fine on version 1909, and it's been working on a non-dongle Bluetooth adapter with 2004. :shrug:
